cluster/cluster.go

```diff
@@ -22,108 +26,104 @@ func validMode(m string) bool {
 }
 
 var (
-	ThisNode *Node
-	Mode     ModeType
-	shutdown = make(chan struct{})
-
-	mu    sync.Mutex
-	peers = make([]*Node, 0)
+	Mode           ModeType
+	Manager        *ClusterManager
+	cfg            *memberlist.Config
+	clusterPrimary = stats.NewBool("cluster.primary")
 )
 
-func Init(name, version string, started time.Time) {
-	ThisNode = &Node{
-		name:          name,
-		started:       started,
-		version:       version,
-		primary:       false,
-		primaryChange: time.Now(),
-		stateChange:   time.Now(),
+func Init(name, version string, started time.Time, scheme string, port int) {
+	Manager = &ClusterManager{
+		Peers: make(map[string]Node),
+		node: Node{
+			Name:          name,
+			ListenPort:    port,
+			ListenScheme:  scheme,
+			Started:       started,
+			Version:       version,
+			Primary:       primary,
+			PrimaryChange: time.Now(),
+			StateChange:   time.Now(),
+			Updated:       time.Now(),
+			local:         true,
+		},
 	}
+	cfg = memberlist.DefaultLANConfig()
+	cfg.BindPort = clusterPort
+	cfg.BindAddr = clusterHost.String()
+	cfg.AdvertisePort = clusterPort
+	cfg.Events = Manager
+	cfg.Delegate = Manager
 }
 
 func Stop() {
-	close(shutdown)
+	Manager.list.Leave(time.Second)
 }
 
 func Start() {
-	go probePeers(probeInterval)
-}
-
-func AddPeer(remoteAddr *url.URL) {
-	mu.Lock()
-	peer := &Node{
-		remoteAddr:  remoteAddr,
-		stateChange: time.Now(),
+	log.Info("Starting cluster on %s:%d", cfg.BindAddr, cfg.BindPort)
+	list, err := memberlist.Create(cfg)
+	if err != nil {
+		log.Fatal(4, "Failed to create memberlist: "+err.Error())
 	}
-	peers = append(peers, peer)
-	go peer.Probe()
-	mu.Unlock()
-}
+	data, err := json.Marshal(Manager.ThisNode())
+	if err != nil {
+		log.Fatal(4, "Failed to marshal ThisNode metadata to json. %s", err.Error())
```

Comment:
shouldn't there be a return or something else happening here? otherwise the next line is going to try to use data which is probably `nil` at this point.